### PR TITLE
ci: Move dist-aarch64-linux to an aarch64 runner

### PR DESCRIPTION
**Pull request by mrkajetanp**
Move the dist-aarch64-linux CI job to an aarch64 runner instead of cross-compiling it from an x86 one. This will make it possible to perform optimisations such as LTO, PGO and BOLT later on.

r? @Kobzol 

try-job: dist-aarch64-linux
try-job: dist-x86_64-linux
try-job: dist-i686-linux

### CODE REVIEW

**Comment by Kobzol** [comment]
First let's see if it works, and what's the time impact :)

@bors try

**Comment by Kobzol** [comment]
The cached duration now seems to be ~1h 20m. Before, it was approximately ~1h 50m, although on a free runner.

**Comment by mrkajetanp** [comment]
Before meaning when it was running on the x86 runner? That would make sense given that it had to build the whole cross-compiling toolchain first I suppose.

**Comment by Kobzol** [comment]
Yeah, before meaning when it was cross-compiling, and also running on a 4-core machine (vs now running on a 8-core machine). The costs of the ARM machine is quite different to us than the previous machine though, so we'll need to discuss this in the infra team.

**Comment by mrkajetanp** [comment]
Fair enough, best of luck! :)

**Review by nikic — CHANGES_REQUESTED** [review]
Our documented glibc baseline for aarch64 is 2.17, while this is switching to ubuntu 22.04, which is glibc 2.35.

**Comment by Kobzol** [comment]
Ah, good point! To clarify, before `crosstool` was used to explicitly configure glibc 2.17 (https://github.com/rust-lang/rust/pull/133809/files#diff-08277b07e8cb8ff6abaa4638229d2a8eb8d14a58590e3ebaafa6d13b3da07081L9), while now we just use a native build on Ubuntu 22.04 (where configuring glibc 2.17 is very painful, at least as far as I'm aware). So we might need to switch e.g. to CentOS, same as we do for the x64 Linux dist build.

**Comment by mrkajetanp** [comment]
> e.g. to CentOS, same as we do for the x64 Linux dist build.

I can switch the build to CentOS, though are we sure that's a good idea given that it's now past its end of life? EOL for the CentOS 7 used in the x86 dist build was June 2024. Afaict all distros' versions that ship 2.17 are now past their EOL.

**Comment by Kobzol** [comment]
The extended LTS is a commercial offering, if I understand correctly, we don't currently make use of any such extended variants.

Created https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Increasing.20glibc.20baseline/near/486974115 to discuss this bump.

**Comment by mrkajetanp** [comment]
> The extended LTS is a commercial offering

From what I understand it's commercial in that a 3rd party gets paid to backport security fixes and such, it's still free to use. Besides it's effectively an opt-in repo so we just use an older official Debian docker image, then if a security fix is needed it's just a matter of adding in the repo and running an update.
https://www.freexian.com/lts/extended/docs/how-to-use-extended-lts/

Whether this is relevant or not is a different question - I imagine not - but it's slightly better than just using a completely dead version I suppose. Especially compared to CentOS where relying on an archive of old builds of a dead distro seems pretty dodgy if we can avoid it.

**Comment by mrkajetanp** [comment]
To clarify, the only concern I have here is that CentOS has been discontinued so I'm wondering whether it's still okay to rely on the repos being up. It's probably good to have at least a migration plan in case they get taken down. If we're confident that they'll stay available for the foreseeable future then there's no problem at all imo.

**Comment by Kobzol** [comment]
Based on https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Increasing.20glibc.20baseline, we're currently OK with using the EOL runner on CI, as we're not really using anything from it statically, we just use it to build the compiler and the stdlib.

**Comment by mrkajetanp** [comment]
Fine by me, let's try the CentOS variant then :)

**Comment by Kobzol** [comment]
@bors try

(The packages should be fixed, but that should be orthogonal to CI duration, it looks like it built even with the x64 packages, lol).

File: src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+RUN yum upgrade -y && \
+    yum install -y \
+      automake \
+      bzip2 \
+      file \
+      gcc \
+      gcc-c++ \
+      git \
+      glibc-devel.i686 \
+      glibc-devel.x86_64 \
```
[the remark this answers is not in the document]
Comment (cuviper) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:23]: I'm really surprised that didn't error in the try build -- I guess yum was permissive about it?

[the remark this answers is not in the document]
Comment (Kobzol) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:23]: Yeah, looks like it just ignored them:
```
2024-12-09T16:54:05.2416487Z #13 79.04 Complete!
2024-12-09T16:54:05.2516652Z #13 79.20 Loaded plugins: ovl
2024-12-09T16:54:05.5352286Z #13 79.48 No package glibc-devel.i686 available.
2024-12-09T16:54:05.6753510Z #13 79.62 No package glibc-devel.x86_64 available.
2024-12-09T16:54:05.8462007Z #13 79.79 No package libstdc++-devel.i686 available.
2024-12-09T16:54:05.9836393Z #13 79.93 No package libstdc++-devel.x86_64 available.
2024-12-09T16:54:06.1639707Z #13 80.11 Package 1:pkgconfig-0.27.1-4.el7.aarch64 already installed and latest version
2024-12-09T16:54:06.3606912Z #13 80.11 Package xz-5.2.2-2.el7_9.aarch64 already installed and latest version
2024-12-09T16:54:06.3607614Z #13 80.17 No package zlib-devel.i686 available.
2024-12-09T16:54:06.3608178Z #13 80.30 No package zlib-devel.x86_64 available.
2024-12-09T16:54:06.5436165Z #13 80.38 Resolving Dependencies
```

[the remark this answers is not in the document]
Comment (mrkajetanp) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:23]: Lol sorry about that, I just used the same list as the x86 build and didn't notice those were there..
Will clean them up tomorrow, no effect on the actual build as you said.

**Comment by Kobzol** [comment]
Ok, 1h 13m, that's pretty good.

**Comment by mrkajetanp** [comment]
Huh, did something change since yesterday?

**Comment by Kobzol** [comment]
Yes, another [PR](https://github.com/rust-lang/rust/pull/133902) was merged that renamed the CI runner. The new name is `job-linux-8c-aarch64`.

**Comment by mrkajetanp** [comment]
Ah yeah I see, thanks - fixed.

File: src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      --build=aarch64-unknown-linux-gnu \
+      --enable-full-tools \
+      --enable-profiler \
+      --enable-sanitizers \
+      --enable-compiler-docs \
+      --set target.aarch64-unknown-linux-gnu.linker=clang \
+      --set target.aarch64-unknown-linux-gnu.ar=/rustroot/bin/llvm-ar \
+      --set target.aarch64-unknown-linux-gnu.ranlib=/rustroot/bin/llvm-ranlib \
+      --set llvm.link-shared=true \
+      --set llvm.thin-lto=false \
```
Comment (Kobzol) [inline src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:85]: Just out of curiosity, why did you disable ThinLTO for LLVM?

File: src/ci/docker/host-aarch64/dist-aarch64-linux/build-clang.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
```
Comment (Kobzol) [inline src/ci/docker/host-aarch64/dist-aarch64-linux/build-clang.sh:1]: Could you please move this script (from here and `dist-x86_64-linux`) to `src/ci/scripts` and parametrize `DLLVM_TARGETS_TO_BUILD` through an environment variable? It's annoying enough to keep the LLVM versions in sync without them being in yet another bash script :)

Same for the GCC/ZSTD build script. And also the `shared.sh` file, it's unnecessarily duplicated here, in `x86_64` and other `dist-XYZ` builderfolders.

Thanks.

**Review by Kobzol — COMMENTED** [review]
Left a few nits regarding duplication, otherwise the infra changes look good to me. Opened a poll on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Switch.20the.20.60dist-aarch64-linux.60.20job.20to.20ARM.20runner) regarding the increase in CI costs by moving to a paid runner.

File: src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile (hunk 2 above)
Comment (mrkajetanp) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:85]: Just to compare the impacts of non-LTO Rust+LLVM with LTO Rust+LLVM. It's enabled in the opt-dist version, should I enable it here as well?

Comment (Kobzol) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/Dockerfile:85]: Yeah, if it works, I would enable it.

File: src/ci/docker/host-aarch64/dist-aarch64-linux/build-clang.sh (hunk 3 above)
Comment (mrkajetanp) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/build-clang.sh:1]: While I'm at it, should I just remove `shared.sh` to `hide-build-output.sh`? Should be clearer as to what it does

Comment (mrkajetanp) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/build-clang.sh:1]: Oh okay this is a bit of a mess. We've got:
- `ci/shared.sh` - a bunch of helpers
- `docker/scripts/shared.sh` - hide_output and retry
- `docker/host-x86_64/dist-x86_64-linux/shared.sh` - slightly different hide_output

Comment (mrkajetanp) [reply src/ci/docker/host-aarch64/dist-aarch64-linux/build-clang.sh:1]: Then I think maybe remove `docker/host-x86_64/dist-x86_64-linux/shared.sh` and just source the one in `docker/scripts/shared.sh`? The functionality should be the same

**Comment by Kobzol** [comment]
Thanks for the cleanup! Let's see.

@bors try

**Comment by mrkajetanp** [comment]
Ah _that's_ what the fixme was for, still needed after all. Should I just set an extra env var just for the 32 bit build and check for it in the script so that we can still share it?

**Comment by Kobzol** [comment]
Yeah, let's do that, it seems like the simplest option.

**Comment by mrkajetanp** [comment]
This fixes it on the machine I have access to, let's give it another try?

**Comment by Kobzol** [comment]
The job had uncached Docker image, and apparently that times out now :/ I'll probably have to increase the timeout again.

@bors try

File: src/ci/docker/scripts/build-clang.sh [hunk 4]
```diff
@@ -34,7 +34,7 @@ hide_output \
       -DCOMPILER_RT_BUILD_XRAY=OFF \
       -DCOMPILER_RT_BUILD_MEMPROF=OFF \
       -DCOMPILER_RT_BUILD_CTX_PROFILE=OFF \
-      -DLLVM_TARGETS_TO_BUILD=X86 \
+      -DLLVM_TARGETS_TO_BUILD=$LLVM_BUILD_TARGETS \
```
Comment (Kobzol) [inline src/ci/docker/scripts/build-clang.sh:37]: nit: I would add `u` to the bash script modifier at the top of this script, just to make sure that this env. var isn't forgotten. Otherwise it could be tricky to debug the CMake failure.

**Comment by davidtwco** [comment]
@mrkajetanp is away for the holidays so I've rebased and fixed the conflict

File: src/ci/docker/scripts/build-clang.sh [hunk 5]
```diff
@@ -5,7 +5,7 @@ set -ex
 source shared.sh
 
 # Try to keep the LLVM version here in sync with src/ci/scripts/install-clang.sh
-LLVM=llvmorg-19.1.0-rc3
+LLVM=llvmorg-19.1.5
```
Comment (Kobzol) [inline src/ci/docker/scripts/build-clang.sh:8]: I would move this update to a separate PR, just to avoid suprirses.

File: src/ci/docker/scripts/build-clang.sh (hunk 4 above)
Comment (davidtwco) [reply src/ci/docker/scripts/build-clang.sh:37]: Fixed

File: src/ci/docker/scripts/build-clang.sh (hunk 5 above)
Comment (davidtwco) [reply src/ci/docker/scripts/build-clang.sh]: Fixed, I'm not sure if @mrkajetanp had changed this deliberately in the `dist-aarch64/build-clang.sh`, so it might be worth confirming it all still works

Comment (mrkajetanp) [reply src/ci/docker/scripts/build-clang.sh]: I bumped it when doing experiments with BOLT, but since we're not enabling bolt for the time being it's not needed as much. Though getting off the RCs if possible is probably good practice in general.

**Comment by Kobzol** [comment]
The duplication is not ideal, but at least it was balanced a bit by the small CI script cleanup :)

@bors r+ rollup=never

**Comment by ionicmc-rs** [comment]
Well it seems this broke something (#134563) related to paging, should this be reverted or fixed.

**Comment by ionicmc-rs** [comment]
Is there a label for something that is `reverted`? if yes, someone add it please
